crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -108,30 +353,123 @@ impl<S: States> Deref for State<S> {
     }
 }
 
+#[derive(Resource, Default, Debug)]
+#[cfg_attr(
+    feature = "bevy_reflect",
+    derive(bevy_reflect::Reflect),
+    reflect(Resource, Default)
+)]
+struct PreviousState<S: States>(S);
```

Comment:
This should probably by pub?

Reply:
I considered this an implementation detail - it only exists in a very limited span, to facilitate some of the OnEnter/OnExit/OnTransition things working correctly. I will go into more detail about why that was necessary in my response to your bigger comments and some of the stuff I think was missed in your branch...

I do think it can be exposed, I just didn't feel that was helpful and didn't want to risk people using it when a better solution could be found later

Reply:
Great, let's leave it private for now. Just make sure to make the methods private though.